resource/traversers/dfu_flexible.cpp

```diff
@@ -0,0 +1,365 @@
+
+        for (Resource r : *slot_resources) {
+            if ((slot_match (u, &r))) {
+                spec = match_kind_t::SLOT_MATCH;
+                pristine = false;
+                ret = slot_resources;
+            }
+        }
+    } else if (match_resources) {
+        spec = match_kind_t::RESOURCE_MATCH;
+        pristine = false;
+        ret = &(match_resources->with);
+    } else {
+        spec = pristine ? match_kind_t::PRISTINE_NONE_MATCH : match_kind_t::NONE_MATCH;
+    }
+
+done:
+    return *ret;
+}
+
+/* Same as above except that lowest is unorder_map */
+int dfu_flexible_t::min_if (subsystem_t subsystem,
+                            resource_type_t type,
+                            unsigned int counts,
+                            std::unordered_map<resource_type_t, int64_t> &lowest)
+{
+    int rc = -1;
+    if (m_match->is_pruning_type (subsystem, type)) {
+        if (lowest.find (type) == lowest.end ())
+            lowest[type] = counts;
+        else if (lowest[type] > counts)
+            lowest[type] = counts;
+        rc = 0;
+    }
+    return rc;
+}
+
+void dfu_flexible_t::prime_jobspec (std::vector<Resource> &resources,
+                                    std::unordered_map<resource_type_t, int64_t> &to_parent)
+{
+    subsystem_t subsystem = m_match->dom_subsystem ();
+    for (auto &resource : resources) {
+        // If the resource is requested as exclusive in the
+        // jobspec, add it to the matcher's exclusive resource
+        // set. This ensures that the full resource set (which
+        // includes shadow resources) is emitted.
+        if (resource.exclusive == Jobspec::tristate_t::TRUE)
+            m_match->add_exclusive_resource_type (resource.type);
+        // Use minimum requirement because you don't want to prune search
+        // as far as a subtree satisfies the minimum requirement
+        accum_if (subsystem, resource.type, resource.count.min, to_parent);
+        prime_jobspec (resource.with, resource.user_data);
+
+        // Or slots should use a minimum of values rather than an accumulation
+        // otherwise possible matches may be filtered out
+        if (resource.type == slot_rt) {
+            for (auto &aggregate : resource.user_data) {
+                min_if (subsystem,
+                        aggregate.first,
+                        resource.count.min * aggregate.second,
+                        to_parent);
+            }
+        } else {
+            for (auto &aggregate : resource.user_data) {
+                accum_if (subsystem,
+                          aggregate.first,
+                          resource.count.min * aggregate.second,
+                          to_parent);
+            }
+        }
+    }
+}
+
+std::tuple<std::map<resource_type_t, int>, int, int> dfu_flexible_t::select_or_config (
+    const std::vector<Resource> &slots,
+    std::map<resource_type_t, int> resource_counts,
```

Comment:
In addition to passing the `resource_counts` parameter by reference (I already mentioned that in another comment), you can use the `const` qualifier here:

```suggestion
    const std::map<resource_type_t, int> &resource_counts,
```